General: Retire legacy deprecated hook code

kraftbj opened the pull request:
WordPress added `do_action_deprecated` and `apply_filters_deprecated`. We've implemented this in our `Jetpack::deprecated_hooks()` function; however, we left our pre-existing legacy custom implementation.

This PR refactors those existing deprecations to the new format.

Additionally, our original implementation returned a PHP notice for trying to access an invalid array offset. That's corrected in [`08d91eb` (#15342)](https://github.com/Automattic/jetpack/pull/15342/commits/08d91ebe63fea0d3a9596eb4de395ab9e484fa70). Please see that commit message for additional context.

#### Changes proposed in this Pull Request:
* Converts legacy deprecation code to use new WordPress-provided functionality.
* Adds tests.
* Resolves PHP array offset error discovered while adding tests.
* PHPCS class.jetpack.php (this was done previously, so this just catches up a few things that snuck in or new coding standards hadn't picked up yet).

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* n/a

#### Testing instructions:
* Add this code to your site, e.g. in debug.php in our docker mu-plugin folder:
```
add_filter( 'jetpack_check_mobile', 'foobug' );
function foobug() { }
```
* Check the error logs and see a deprecation notice.
* `yarn docker:phpunit --filter=test_deprecated

#### Proposed changelog entry for your changes:
* General: retire legacy implementation of throwing deprecation notices for old hooks.

kraftbj committed:
Report version of Jetpack that deprecated a function

mdbitz reviewed (commented):
Incomplete comment, sigh

This is definitely an improvement as we are showing the version for what is known. However should the default be something other than empty ?

perhaps "Jetpack" to highlight that they are trying to reference functionality in Jetpack that doesn't exist.

`PHP Deprecated: jetpack_holiday_snowing used for foobug is <strong>deprecated</strong> since version with no alternative available. in /var/www/html/wp-includes/functions.php on line 4723`

kraftbj committed:
Provide default version of 'Jetpack'

kraftbj committed:
Deprecated Hooks: Add initial attempt of a test

kraftbj committed:
Migrate previous custom solution array to WP version

kraftbj committed:
PHPCS -- done previously, so cleaning up a few things that slipped in over time

kraftbj committed:
Improve test

kraftbj committed:
Remove old legacy implementation

kraftbj committed:
Fix invalid offset error in original code.

`apply_filters_deprecated`'s second argument is an array of the items passed to the original filter. Filters expect an original value, so does `apply_filters_deprecated`. To allow for multiple values (e.g. `apply_filters( $name, $arg1, $arg2 )`, apply_filters_deprecated handle this via `apply_filters_deprecated( $name, $array )`, where `$array = array( $arg1, $arg2 ) )`.

The WP code returns the value of `$array[0]` ( see `wp-includes/plugin.php L248`, so we have to set _something_ in the array.

kraftbj commented:
This saw a pretty big update since a lot changed since I last touched this back in April. The PR title and description has been updated to reflect the current patch.

anomiex reviewed (approved):
Seems ok, and the unit tests pass.

I see the linter flagged a few more PHPCS things you might clean up if you want, since you already cleaned up a bunch of them.

kraftbj commented:
Thanks, @anomiex. I'm going to merge without those PHPCS edits. I just did PHPCBF and there's... a lot... to bring it up for full-compliance.
